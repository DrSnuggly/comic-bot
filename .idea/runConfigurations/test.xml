<component name="ProjectRunConfigurationManager">
  <configuration default="false" name="test" type="CompoundRunConfigurationType">
    <toRun name="debugger" type="ChromiumRemoteDebugType" />
    <toRun name="All Tests" type="JavaScriptTestRunnerVitest" />
    <method v="2" />
  </configuration>
</component>