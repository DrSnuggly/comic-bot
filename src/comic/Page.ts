import { decodeHTML } from "entities"
import type { ComicData } from "../schema"
import { ComicError } from "./ComicError"

export type PageData = {
  imageUri: string
  altText?: string | undefined
}

export class PageError extends ComicError {}

/**
 * Content handler for {@link HTMLRewriter}.
 */
export class Page implements PageData {
  imageUri: string
  altText: string | undefined

  /**
   * @param comic - the target comic.
   * @param rewriter - the instantiated {@link HTMLRewriter} object.
   * @param data - the parsed page data.
   */
  protected constructor(
    public comic: ComicData,
    protected rewriter: HTMLRewriter,
    data: PageData,
  ) {
    this.imageUri = data.imageUri
    this.altText = data.altText
  }

  /**
   * Fetch and parse the comic.
   * @param url - the target comic page.
   * @param comic - the target comic.
   * @param rewriter - the instantiated {@link HTMLRewriter} object.
   */
  static async fromUrl(
    url: string,
    comic: ComicData,
    rewriter: HTMLRewriter,
  ): Promise<Page> {
    const response = await fetch(url)
    return Page.fromResponse(response, comic, rewriter)
  }

  /**
   * Fetch and parse the comic.
   * @param response - the target comic page response.
   * @param comic - the target comic.
   * @param rewriter - the instantiated {@link HTMLRewriter} object.
   * @throws PageError if bad response or unable to parse.
   */
  static async fromResponse(
    response: Response,
    comic: ComicData,
    rewriter: HTMLRewriter,
  ): Promise<Page> {
    if (!response.ok) {
      throw new PageError(
        comic,
        `received ${response.status} when retrieving ${response.url}: ${await response.text()}`,
      )
    }

    let imageUri: string | undefined
    let altText: string | undefined
    const handler: HTMLRewriterElementContentHandlers = {
      element(element) {
        // Get comic image, if not already found.
        if (imageUri === undefined) {
          const src = element.getAttribute("src")
          if (src) imageUri = src
        }

        // Get comic alt text, if not already found.
        if (altText === undefined) {
          const title = Page.normalizeAltText(element.getAttribute("title"))
          if (title !== undefined) altText = title
        }
      },
      text(element) {
        // Get comic alt text, if not already found.
        if (altText === undefined) {
          const text = Page.normalizeAltText(element.text)
          if (text !== undefined) altText = text
        }
      },
    }

    let parser = rewriter.on(comic.imageSelector, handler)
    if (comic.altSelector) {
      parser = parser.on(comic.altSelector, handler)
    }
    await parser.transform(response).text()
    if (imageUri === undefined) {
      throw new PageError(comic, "could not find image URI")
    }

    return new Page(comic, rewriter, { imageUri, altText })
  }

  /**
   * Normalize input into the expected value.
   * @param input - the input to normalize.
   */
  static normalizeAltText(input: unknown): string | undefined {
    if (!input || typeof input !== "string") return undefined
    const parsed = decodeHTML(input).trim()
    if (!parsed) return undefined
    return parsed
  }
}
