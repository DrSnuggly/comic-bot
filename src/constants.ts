export const KV_INDEX_KEY = "index"
