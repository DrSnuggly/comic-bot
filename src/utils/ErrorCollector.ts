/**
 * Guard for non-string iterables.
 * @param obj - the item to test.
 */
function isNonStringIterable(
  obj: unknown,
): obj is Exclude<Iterable<unknown>, string> {
  if (typeof obj === "string") return false
  return typeof (obj as Iterable<unknown>)[Symbol.iterator] === "function"
}

export class ErrorCollector extends Set<unknown> {
  override add(e: unknown): this {
    if (!isNonStringIterable(e)) return super.add(e)
    for (const item of e) super.add(item)
    return this
  }

  /**
   * Add any errors from `await Promise.allSettled()`.
   * @param settled - the settled promises array.
   */
  addSettled(settled: PromiseSettledResult<unknown>[]): this {
    for (const item of settled.filter((item) => item.status === "rejected")) {
      this.add(item.reason)
    }
    return this
  }

  /**
   * Assert that no items have been collected.
   * @throws unknown[] for the collected errors.
   */
  assertEmpty(): void {
    if (this.size) throw Array.from(this)
  }
}
