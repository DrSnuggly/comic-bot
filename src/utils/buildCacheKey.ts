import type { ComicData } from "../schema"

/**
 * Generate a cache key for storing and retrieving last update times.
 * @param url - a webhook URL.
 * @param comic - the target comic.
 */
export function buildCacheKey(url: string, comic: ComicData): string {
  return `${comic.rssUrl}|${url}`
}
