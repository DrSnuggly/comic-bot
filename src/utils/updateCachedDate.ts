import { env } from "cloudflare:workers"
import type { ComicData } from "../schema"
import { buildCacheKey } from "./buildCacheKey"

/**
 * Update the cached date for a given comic and webhook.
 * @param url - a webhook URL.
 * @param comic - the target comic.
 */
export function updateCachedDate(url: string, comic: ComicData): Promise<void> {
  return env.KV.put(buildCacheKey(url, comic), new Date().toISOString())
}
