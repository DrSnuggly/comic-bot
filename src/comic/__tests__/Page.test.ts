import { env, fetchMock, type Interceptable } from "cloudflare:test"
import { comicData, comicUrl, origin } from "@tests/utils/constants"
import { fetchAsset } from "@tests/utils/fetchAsset"
import type { ComicData } from "../../schema"
import { Page, type PageData, PageError } from "../Page"

let rewriter: HTMLRewriter
let mockPool: Interceptable
beforeAll(() => {
  rewriter = new HTMLRewriter()
  fetchMock.activate()
  fetchMock.disableNetConnect()
  mockPool = fetchMock.get(origin)
})
afterEach(() => fetchMock.assertNoPendingInterceptors())

describe("validation", () => {
  it("should reject if response not ok", async () => {
    mockPool.intercept({ path: comicUrl.pathname }).reply(400)
    const pagePromise = Page.fromUrl(comicUrl.href, comicData, rewriter)

    await expect(pagePromise).rejects.instanceof(PageError)
  })

  it("should reject if comic not found", async () => {
    const assetPath = "comic-pages/no-comic.html"

    const asset = await fetchAsset(assetPath)
    const pagePromise = Page.fromResponse(asset, comicData, rewriter)

    await expect(pagePromise).rejects.instanceof(PageError)
  })
})

describe("parsing", () => {
  it.for([
    { pageAsset: "inline-alt-text.html" },
    { pageAsset: "inline-nullish-alt-text.html" },
    { pageAsset: "no-alt-text.html" },
    { pageAsset: "no-alt-text.html", altSelector: "#does-not-exist" },
    { pageAsset: "separate-alt-text.html", altSelector: "#alt-text" },
  ] satisfies (Pick<ComicData, "altSelector"> & { pageAsset: string })[])(
    "should parse: $pageAsset",
    async ({ pageAsset, ...rest }) => {
      const assetPath = `comic-pages/${pageAsset}`
      const testComic = { ...comicData, ...rest }

      const asset = await env.ASSETS.fetch(`${origin}/${assetPath}`)
      const pagePromise = Page.fromResponse(asset, testComic, rewriter).then(
        ({ imageUri, altText }) => ({ imageUri, altText }) satisfies PageData,
      )

      await expect(pagePromise).resolves.toMatchSnapshot()
    },
  )
})
