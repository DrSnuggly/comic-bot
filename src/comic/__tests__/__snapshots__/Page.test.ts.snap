// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parsing > should parse: 'inline-alt-text.html' 1`] = `
{
  "altText": "Alt text.",
  "imageUri": "https://localhost/image.png",
}
`;

exports[`parsing > should parse: 'inline-nullish-alt-text.html' 1`] = `
{
  "altText": undefined,
  "imageUri": "https://localhost/image.png",
}
`;

exports[`parsing > should parse: 'no-alt-text.html' 1`] = `
{
  "altText": undefined,
  "imageUri": "https://localhost/image.png",
}
`;

exports[`parsing > should parse: 'no-alt-text.html' 2`] = `
{
  "altText": undefined,
  "imageUri": "https://localhost/image.png",
}
`;

exports[`parsing > should parse: 'separate-alt-text.html' 1`] = `
{
  "altText": "Alt text.",
  "imageUri": "https://localhost/image.png",
}
`;

exports[`should parse: 'inline-alt-text.html' 1`] = `
{
  "altText": "Alt text.",
  "imageUri": "https://localhost/image.png",
}
`;

exports[`should parse: 'no-alt-text.html' 1`] = `
{
  "altText": undefined,
  "imageUri": "https://localhost/image.png",
}
`;

exports[`should parse: 'no-alt-text.html' 2`] = `
{
  "altText": undefined,
  "imageUri": "https://localhost/image.png",
}
`;

exports[`should parse: 'separate-alt-text.html' 1`] = `
{
  "altText": "Alt text.",
  "imageUri": "https://localhost/image.png",
}
`;
